app/main/views/frameworks.py

```diff
@@ -566,3 +593,178 @@ def upload_framework_agreement(framework_slug):
+        # No file chosen for upload and file already exists on s3 so can use existing and progress
+        if not request.files['signature_page'].filename and signature_page:
+            return redirect(url_for(".contract_review", framework_slug=framework_slug))
+
+        if not file_is_image(request.files['signature_page']) and not file_is_pdf(request.files['signature_page']):
+            upload_error = "The file must be a PDF, JPG or PNG"
+        elif not file_is_less_than_5mb(request.files['signature_page']):
+            upload_error = "The file must be less than 5MB"
+        elif file_is_empty(request.files['signature_page']):
+            upload_error = "The file must not be empty"
+
+        if not upload_error:
+            upload_path = get_agreement_document_path(
+                framework_slug,
+                current_user.supplier_id,
+                '{}{}'.format(SIGNED_AGREEMENT_PREFIX, get_extension(request.files['signature_page'].filename))
+            )
+            agreements_bucket.save(
+                upload_path,
+                request.files['signature_page'],
+                acl='private'
+            )
+
+            session['signature_page'] = request.files['signature_page'].filename
+
+            return redirect(url_for(".contract_review", framework_slug=framework_slug))
+
+    return render_template(
+        "frameworks/signature_upload.html",
+        framework=framework,
+        signature_page=signature_page,
+        upload_error=upload_error,
+    ), 400 if upload_error else 200
+
+
+@main.route('/frameworks/<framework_slug>/contract-review', methods=['GET', 'POST'])
+@login_required
+def contract_review(framework_slug):
+    framework = get_framework(data_api_client, framework_slug)
+    supplier_framework = return_supplier_framework_info_if_on_framework_or_abort(data_api_client, framework_slug)
+    agreements_bucket = s3.S3(current_app.config['DM_AGREEMENTS_BUCKET'])
+    signature_page = get_most_recently_uploaded_agreement_file_or_none(agreements_bucket, framework_slug)
+
+    # if supplier_framework doesn't have a name or a role or the agreement file, then redirect to first page
+    if not (
+        supplier_framework['agreementDetails'] and
+        supplier_framework['agreementDetails'].get('signerName') and
+        supplier_framework['agreementDetails'].get('signerRole') and
+        signature_page
+    ):
+        return render_template("errors/500.html"), 400
+
+    form = ContractReviewForm()
+    form_errors = None
+
+    if request.method == 'POST':
+        if form.validate_on_submit():
+            data_api_client.register_framework_agreement_returned(
+                current_user.supplier_id, framework_slug, current_user.email_address, current_user.id
+            )
+
+            email_recipients = [supplier_framework['declaration']['primaryContactEmail']]
+            if supplier_framework['declaration']['primaryContactEmail'] != current_user.email_address:
```

Comment:
considering de-case-sensitivizing this?

in theory the _pre-domain part_ of an email can be case sensitive, but in practise almost no mail services treat differently-cased mailboxes differently.